numba/np/arraymath.py

```diff
@@ -3674,107 +3674,177 @@ def bincount_impl(a, weights=None, minlength=0):
-                # mid is too low => go up
-                lo = mid + 1
+            return False
+
+    else:
+        if np.isnan(b.real):
+            return True
+        else:
+            if np.isnan(a.imag):
+                if np.isnan(b.imag):
+                    return a.real <= b.real
+                else:
+                    return False
             else:
-                # mid is too high, or is a NaN => go down
-                hi = mid
-        return lo
-    return searchsorted_inner
+                if np.isnan(b.imag):
+                    return True
+                else:
+                    if a.real < b.real:
+                        return True
+                    elif a.real == b.real:
+                        return a.imag <= b.imag
+                    return False
+
+
+def _searchsorted(func_1, func_2):
+    # a facsimile of:
+    # https://github.com/numpy/numpy/blob/4f84d719657eb455a35fcdf9e75b83eb1f97024a/numpy/core/src/npysort/binsearch.cpp#L61  # noqa: E501
+
+    def impl(a, v):
+        min_idx = 0
+        max_idx = len(a)
 
+        out = np.empty(v.size, np.intp)
 
-_lt = less_than
-_le = register_jitable(lambda x, y: x <= y)
-_searchsorted_left = register_jitable(_searchsorted(_lt))
-_searchsorted_right = register_jitable(_searchsorted(_le))
+        last_key_val = v.flat[0]
+
+        for i in range(v.size):
+            key_val = v.flat[i]
+
+            if func_1(last_key_val, key_val):
+                max_idx = len(a)
+            else:
+                min_idx = 0
+                if max_idx < len(a):
+                    max_idx += 1
+                else:
+                    max_idx = len(a)
+
+            last_key_val = key_val
+
+            while min_idx < max_idx:
+                # to avoid overflow
+                mid_idx = min_idx + ((max_idx - min_idx) >> 1)
+
+                mid_val = a[mid_idx]
+
+                if func_2(mid_val, key_val):
+                    min_idx = mid_idx + 1
+                else:
+                    max_idx = mid_idx
+
+            out[i] = min_idx
+
+        return out.reshape(v.shape)
+
+    return impl
+
+
+VALID_SEARCHSORTED_SIDES = frozenset({'left', 'right'})
+
+
+def bind_operators(side, lt, le, use_version_gate=False):
+    if side == 'left':
+        _impl = _searchsorted(lt, lt)
+    else:
+        if use_version_gate and numpy_version < (1, 23):
+            # change in behaviour for floats and complex
+            # introduced by:
+            # https://github.com/numpy/numpy/pull/21867
+            _impl = _searchsorted(lt, le)
+        else:
+            _impl = _searchsorted(le, le)
+
+    return _impl
+
+
+def make_searchsorted_implementation(np_dtype, side):
+    assert side in VALID_SEARCHSORTED_SIDES
+
+    if np.issubdtype(np_dtype, np.floating):
+        _impl = bind_operators(
+            side,
+            less_than_float,
+            less_than_or_equal_float,
+            use_version_gate=True
+        )
```

Comment:
Could we reduce code duplication here and leverage the use of `isinstance` to avoid having a `less_than_float` and `less_than_complex`? For instance, the `less_than_or_equal` function could be enhanced to support floats with the following addition:

```python
@register_jitable
def less_than_or_equal(a, b):
    if isinstance(b, float):
        if np.isnan(b):
            return True
    return a <= b
```

For the following to work, you need to update the `supported_var_ty` variable in `builtins.py` to add two new entries:

```python
    supported_var_ty = (types.Number, types.Bytes, types.RangeType,
                        types.DictType, types.LiteralStrKeyDict, types.List,
                        types.ListType, types.Tuple, types.UniTuple, types.Set,
                        types.Function, types.ClassType, types.UnicodeType,
                        types.ClassInstanceType, types.NoneType, types.Array,
                        types.Boolean, types.UnicodeCharSeq)  # <----
```

Reply:
Thanks very much for the review and feedback, @guilhermeleobas.

In my local branch, I have put something together which avoids separations like `less_than_float` and `less_than_complex`, but it has ended up looking a bit odd and might conceivably have performance / other implications. 

This probably means I've not interpreted your feedback correctly - nonetheless, I am happy to push it for review if that's useful.

Reply:
Further to the above, I've pushed my branch - let me know what you think in due course.